### ISSUE TITLE
Source Generation silently crashes when using wrong compiler version

### ISSUE THREAD

**Issue by tarekgh** [dotnet/roslyn#66681]
**Version Used**: 
`4.3.0-3.22470.13 (80a8ce8d)`

** Details

We got a project compiling using `2022 VS v17.3.6` which using the compiler version `4.3.0-3.22470.13 (80a8ce8d)`. This project using [logging source generation code](https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator#basic-usage). The logging source generator needs at least the compiler version `4.4`. It happens this project is disabling the warning `CS9057`. Building such project produces error like `C:\scratch\webApp\Program.cs(35,32): error CS8795: Partial method 'Log.Hello(ILogger, string)' must have an implementation part because it has accessibility modifiers. [C:\scratch\webApp\webApp.csproj]`. This error gives hint but not enough to tell that the source generator crashed because of using wrong compiler version. This made the problem diagnostics not easy to know what is going on.

**Steps to Reproduce**:

create any project using `4.3.0-3.22470.13 (80a8ce8d)` compiler version and inside this project use the [logging source generation code](https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator#basic-usage). also try to disable `CS9057` in that project.

**Expected Behavior**:

Compiler should produce a warning when the source generator crashes even if `CS9057` is disabled.  

**Actual Behavior**:

No warning is produced and will get a compilation error `C:\scratch\webApp\Program.cs(35,32): error CS8795: Partial method 'Log.Hello(ILogger, string)' must have an implementation part because it has accessibility modifiers. [C:\scratch\webApp\webApp.csproj]` which is not clear what is exactly the problem especially compiling the exact same project in another environment using compiler version `4.4` will succeed.

**Comment by ericstj** [comment]
So the compiler did emit a warning, but the user suppressed it.  
```
CSC : warning CS9057: The analyzer assembly 'C:\Program Files\dotnet\packs\Microsoft.AspNetCore.App.Ref\7.0.2\analyzers\dotnet\roslyn4.4\cs\Microsoft.Extensions.Logging.Generators.dll' references version '4.4.0.0' of the compiler, which is newer than the currently running version '4.3.0.0'.
```
After suppressing this, nothing appears in the log, not even an informational diagnostic in the binlog.  Either the warning shouldn't be suppressible (error), or it should still leave a trace after suppressing it.

Perhaps if the user had other ways of controlling analyzers (enabling/disabling) then this could be an error resolved by disabling the analyzer.

**Comment by chsienki** [comment]
So in this case the generator isn't crashing. It never gets loaded in the first place because we know the version doesn't match (hence the CS9057). If the user explicitly suppresses that message, I don't see what else we can do?

**Comment by chsienki** [comment]
We could possibly alter the message to make it clearer that whatever is contained in that assembly won't get run?

**Comment by RikkiGibson** [comment]
More context can be found in #61252.

**Comment by RikkiGibson** [comment]
> it should still leave a trace after suppressing it.

I'm not opposed to this. I wonder what the most expedient way to do this would be. For example, if we emitted a hidden diagnostic as well as a warning, would we be guaranteed to have *something* show up in the binlog. It feels like this would be applicable for any analyzer load failure warning.

Also, I wonder if there is any expedient way we could introduce granular suppressions of analyzer load failure warnings. For example, suppressing warnings for an analyzer assembly with a particular simple name.

**Comment by jaredpar** [comment]
> Either the warning shouldn't be suppressible (error), 

The problem with making it an error is there are scenarios where failing to load an analyzer / generator will result in a successful compilation. Analyzers in particular are very often _optional_. 

When we initially designed this diagnostic it was not creating a new scenario. Rather it was taking an existing scenario which may or may not block compilation and giving it a much more deterministic value. For such scenarios warnings are preferred to minimize chances of creating a blocking issue. 

> or it should still leave a trace after suppressing it.

Disagree. The compiler gave a warning that details thee problem. If customers suppress that diagnostic then they're subverting our attempt to warn them about the situation.
